CodeU_Assignement_3/src/codeu_assignement_3/KemWordSearch.java

```diff
@@ -1,16 +1,138 @@
 package codeu_assignement_3;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
 /**
  *
  * @author Karine
  */
 public class KemWordSearch 
 {
-    private KemDictionary m_dictionary;
+    /*
+    Input: the number of rows, number of columns, 
+           a 2-dimensional array of characters (of the native char data type)
+           a dictionary
+    
+    Process: 
     
-    public KemWordSearch (KemDictionary dictionary) 
+    Output: return the set of all-words found
+    */
+    String[] wordSearch(int row, int col, char[][] grid, KemDictionary dictionary)
     {
-        m_dictionary = dictionary; 
-        /* To search word must init first with the dictionary we wish to search in */
+        /* Corner cases */ 
+        if (grid == null) {
+            return null;
+        }
+        if (dictionary == null)
+        {
+            return null;
+        }
+        
+        ArrayList<String> ret_temp = new ArrayList<>();
+
+        /* Search for all possible strings from the grid */
+        for (int r = 0 ; r < row; r++)
+        {
+            for (int c = 0 ; c < row; c++)
+            {
+                /* Search for all the words from a specific location */
+                String[] ret_inner 
+                        = wordSearchFromPossition(row,col,grid,dictionary,r,c,"");
+                
+                /* If there is something to add, add it */
+                if (ret_inner != null)
+                {
+                    ret_temp.addAll(Arrays.asList(ret_inner));
+                }
+            }
+        }
+        
+        /* Return the set from any start point */
+        return (ret_temp.isEmpty()) ? null : (ret_temp

```

Comment:
Why return null for an empty collection? An empty collection is easy to test for and means less special-case code needed for our caller.

Moreover, you could just use a Set to hold the words, and then you don't need to remove duplicates. This has the potential to save memory if the algorithm generates a large number of duplicates, as you will automatically de-duplicate as you go.